### PR TITLE
Convert DownloadRequest resource/controller to kubebuilder 

### PR DESCRIPTION
**Pull request by carlisia**
This closes https://github.com/vmware-tanzu/velero/issues/2656.

### CODE REVIEW

**Comment by nrb** [comment]
@carlisia I did a quick pass over the code, but is there anything in particular preventing it from moving out of draft right now?

**Comment by carlisia** [comment]
> @carlisia I did a quick pass over the code, but is there anything in particular preventing it from moving out of draft right now?

Yes, I need to fix this test: https://github.com/vmware-tanzu/velero/pull/3004/files#diff-6d5dc1afa6af05350b9f38368bde033f93c478b86304922796d2e6eedcbf1336

**Comment by carlisia** [comment]
Note: @nrb agreed that the test pkg/cmd/util/downloadrequest/downloadrequest_test.go should be removed. To be sure, the controller has thorough testing.

File: config/rbac/role.yaml
Comment (nrb) [inline config/rbac/role.yaml:40]: Since we didn't implement a `watch` in the controller, should we remove this line?

File: pkg/cmd/util/downloadrequest/downloadrequest.go
Comment (nrb) [inline pkg/cmd/util/downloadrequest/downloadrequest.go:62]: Pretty sure we can actually remove these checks now. However, let's not in this PR :)

File: pkg/cmd/util/downloadrequest/downloadrequest.go
Comment (nrb) [inline pkg/cmd/util/downloadrequest/downloadrequest.go:44]: I don't see the `timeout` argument used anywhere in the new code - am I missing it?

**Review by nrb — CHANGES_REQUESTED** [review]
A couple questions, and I think one of the parameters passed in to the `Stream` function is no longer used now.

File: pkg/cmd/util/downloadrequest/downloadrequest.go
Comment (zubron) [reply pkg/cmd/util/downloadrequest/downloadrequest.go:44]: It's used on line 105.

File: config/rbac/role.yaml
Comment (carlisia) [reply config/rbac/role.yaml:40]: The very first bullet on this issue is to clean this up: https://github.com/vmware-tanzu/velero/issues/2884

I'll be trying to avoid this from being generated too.

File: pkg/cmd/util/downloadrequest/downloadrequest.go
Comment (carlisia) [reply pkg/cmd/util/downloadrequest/downloadrequest.go:62]: Created an issue for this: https://github.com/vmware-tanzu/velero/issues/3126

**Comment by carlisia** [comment]
Comments so far have been addressed, ready for review.

**Review by nrb — COMMENTED** [review]
Dismissing request for changes, but looks like a rebase is necessary.

**Comment by carlisia** [comment]
Rebased, ready for review.

**Review by nrb — DISMISSED** [review]

**Comment by nrb** [comment]
@vmware-tanzu/velero-maintainers Please queue this for review before break

File: pkg/controller/download_request_controller.go
Comment (dsu-igeek) [inline pkg/controller/download_request_controller.go:129]: Should we use errors.WithStack() here like above?

**Review by dsu-igeek — APPROVED** [review]
I just had the places where I think we want to use WithStack or Wrap on the returned error, otherwise looks good.

File: pkg/controller/download_request_controller_test.go
Comment (zubron) [inline pkg/controller/download_request_controller_test.go:48]: Can these be removed if we're not doing anything in setup/teardown?

File: pkg/controller/download_request_controller_test.go
Comment (zubron) [inline pkg/controller/download_request_controller_test.go:55]: I think if we're going to use Ginkgo here, we should be using the `DescribeTable` function for these tests as you did in #3180. I think without that, all these tests are treated as one spec and you don't get to see the name of a test if it fails.

File: pkg/controller/download_request_controller_test.go
Comment (zubron) [inline pkg/controller/download_request_controller_test.go:193]: In all the test cases, `tc.downloadRequest` is not nil and it seems like it's necessary that it isn't nil. If you wanted to be really strict, you could fail the test early if it's nil?

**Review by zubron — CHANGES_REQUESTED** [review]
I know you've shown me some of the work to use kubebuilder/controller runtime before, but seeing the changes in a PR make it really clear how much nicer this approach is :smile: 

Most of the comments I've left are cleanup/suggestions and could be addressed in a follow up. The only thing I'd really like to check before merging is the change to the requeue in the BSL controller. I don't know if that's an intentional change, and if so, what is the impact.

File: pkg/controller/download_request_controller.go
Comment (carlisia) [reply pkg/controller/download_request_controller.go:70]: The reason for returning some errors with stack and others not is documented here: https://velero.io/docs/v1.5/code-standards/#code.

I think there are some in this code that should be returning with Stack but are not. I'll revise, thank you.

File: pkg/controller/backup_storage_location_controller.go
Comment (carlisia) [reply pkg/controller/backup_storage_location_controller.go:63]: It was a fix because whenever an error is returned, the request gets requeued, so it was redundant.

File: pkg/controller/download_request_controller_test.go
Comment (carlisia) [reply pkg/controller/download_request_controller_test.go:48]: Yes, good catch, thank you! I'll remove it from #3180 too as well as for other merged controllers.

File: pkg/controller/download_request_controller_test.go
Comment (carlisia) [reply pkg/controller/download_request_controller_test.go:55]: Done.

File: pkg/controller/backup_storage_location_controller.go
Comment (zubron) [reply pkg/controller/backup_storage_location_controller.go:63]: Thanks for explanation!

File: pkg/controller/download_request_controller_test.go
Comment (carlisia) [reply pkg/controller/download_request_controller_test.go]: Yes, good point. Fixed.

**Comment by carlisia** [comment]
Hey @vmware-tanzu/velero-maintainers, thanks for the reviews and sorry for the long turn around.

This is ready for review again.

File: pkg/cmd/util/downloadrequest/downloadrequest.go
Comment (nrb) [inline pkg/cmd/util/downloadrequest/downloadrequest.go:2]: 2021 now.

File: pkg/controller/download_request_controller.go
Comment (nrb) [inline pkg/controller/download_request_controller.go:127]: If we never find a DownloadURL here, will it just constantly re-queue?

Maybe we should check for expiration at the beginning, regardless of whether the downloadrequest.Status.Phase field is processed, so that erroneous requests don't sit and spin in the controller forever, filling up the logs with errors. Since DownloadRequests are usually a synchronous activity where someone tries to download a thing at the command line, I'm not sure if it's worth continuing to try past the TTL. Does that sound reasonable to others?

**Review by nrb — CHANGES_REQUESTED** [review]
Unfortunately copyright year changed :(. We do have safe legal ground to remove the copyright year, though, so if you'd rather do that for your changed files, 👍  though we should probably scan through and remove them all at once.

Also a question about requeueing failed requests. I asked it specifically on the DownloadURL, but I think it applies to all of the types.

File: pkg/controller/download_request_controller.go
Comment (zubron) [reply pkg/controller/download_request_controller.go:127]: Right now it looks like the expiration is only set once a DownloadURL has been received (which happens in the case of success, we have a DownloadURL which expires after TTL). Are you suggesting that we should instead set the expiration at the point where we **create** the DownloadRequest, and then the expiration of the request would be tied to time since creation, rather than the expiration time of the resulting URL?

Comment (carlisia) [reply pkg/controller/download_request_controller.go:143]: Addressed this as per Nolan's suggestion.

**Comment by carlisia** [comment]
I am manually testing this and getting the error below, will add a comment when it's working and ready for review.

```
An error occurred: Get "http://minio.velero.svc:9000/velero/backups/test1/test1.tar.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20210208%2Fminio%2Fs3%2Faws4_request&X-Amz-Date=20210208T145945Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=1b5aafa73c614741cb20f3cc098e5257e1c19e6111599c53ff8f9fb21ffba7a4": dial tcp 10.195.47.241:9000: connect: connection refused
```

**Comment by zubron** [comment]
@carlisia I've seen that error before and it's caused by MinIO not being accessible outside the cluster. I managed to fix it by doing a `kubectl port-forward <minio-pod> 9000` and adding `publicUrl: http://localhost:9000` to the BSL config. I found that in the AWS plugin docs: https://github.com/vmware-tanzu/velero-plugin-for-aws/blob/main/backupstoragelocation.md

I think we need to update our docs for Tilt to include this info.

**Comment by carlisia** [comment]
Oh wow, thank you!

**Comment by carlisia** [comment]
@zubron please paste your answer here: https://github.com/vmware-tanzu/velero/discussions/3420

I can update the docs later.

**Comment by carlisia** [comment]
@vmware-tanzu/velero-owners this release blocker is ready for review. 

With Bridget's suggestion I was able to complete a manual test. 🎉

**Comment by nrb** [comment]
@carlisia Awesome! Looks like it'll need a rebase

**Comment by carlisia** [comment]
@vmware-tanzu/velero-owners this release blocker is ready for review.

File: pkg/apis/velero/v1/download_request_types.go
Comment (jenting) [inline pkg/apis/velero/v1/download_request_types.go:2]: nit: Remove the year

File: pkg/apis/velero/v1/download_request_types.go
Comment (jenting) [inline pkg/apis/velero/v1/download_request_types.go:80]: typo: runtime

File: pkg/apis/velero/v1/download_request_types.go
Comment (jenting) [inline pkg/apis/velero/v1/download_request_types.go:108]: typo: runtime

File: pkg/builder/download_request_builder.go
Comment (jenting) [inline pkg/builder/download_request_builder.go:57]: This should be the Target function description.

**Comment by nrb** [comment]
Looks like we have conflicts after the colorized output PR merged.

**Comment by carlisia** [comment]
Ready for review.

**Review by nrb — CHANGES_REQUESTED** [review]
I think this still needs a little work.

My original intention was to add a new case where DownloadRequests were not reaching a Processed phase, but reached the expiration time. These requests would just constantly requeue in the old code, because there was no terminal state for DownloadRequests never successfully got a DownloadURL. Eventually, these could pile up.

We should also reach a terminal state when the DownloadRequest is moved to the Processed phase, which is what we do now.

So in total, there should be 2 ways for a single DownloadRequest to be deleted in the reconciliation loop:

1) "Now" is after the expiration time; we've never reached the Processed state, but also don't want to queue infinitely
2) We've reach the Processed state, and can clean it up.

File: pkg/controller/download_request_controller_test.go
Comment (zubron) [inline pkg/controller/download_request_controller_test.go:247]: Further to @nrb's comments, we should add some more test cases here to show that if a request has phase 'New', or is empty, that we should still delete it if it has expired. I was going through some of the test logic and it looks like it only applies expiration logic in the case where the status is Processed.

**Comment by zubron** [comment]
> 2. We've reach the Processed state, and can clean it up.

Just to clarify - we currently delete the request if it's Processed *and* it's expired. I'm not totally clear on the logic of the controllers and how the download request streaming works but if the request is moved to Processed, it will be requeued and would then be deleted on the next reconciliation loop. It might still be in use by the logs streaming and shouldn't be deleted until after its expiration.

**Comment by nrb** [comment]
> Just to clarify - we currently delete the request if it's Processed and it's expired. 

Thanks, you're right. I was a little hand-wavy there as I was focusing on the current loop iteration. But you're right - it'll requeue until it's both Processed _and_ expired.

**Comment by carlisia** [comment]
Addressed the corrections (thank you1).

Note: I don't know what happened but needed to do a rebase, and the actual commit was "eaten", but the changes were pushed.

Ready for review.

**Review by nrb — APPROVED** [review]

File: pkg/controller/download_request_controller_test.go
Comment (zubron) [inline pkg/controller/download_request_controller_test.go:262]: Like the previous test where the expired `DownloadRequest` was deleted, we shouldn't expect an error here. There are a few points in the main test code where we only set the expiration on the `DownloadRequest` when it is `Processed` when it should be applied to any expired request.

File: pkg/controller/download_request_controller_test.go
Comment (zubron) [inline pkg/controller/download_request_controller_test.go:85]: I think this whole outer check can be removed as we need to set the expiration for all requests.

File: pkg/controller/download_request_controller_test.go
Comment (zubron) [inline pkg/controller/download_request_controller_test.go:145]: If `test.expired` is true, then we should always expect the err to be a `NotFound` error.

**Review by zubron — CHANGES_REQUESTED** [review]
Thanks, @carlisia! There are just a couple of changes that need to be made in one of the test cases but it's good otherwise :+1:

File: pkg/controller/download_request_controller_test.go
Comment (carlisia) [reply pkg/controller/download_request_controller_test.go:85]: The expiration gets set only in the same block of code where we update the request status to `processed` (https://github.com/vmware-tanzu/velero/blob/1c536aced492862ed61842315921ab6ccfe62887/pkg/controller/download_request_controller.go#L113). 

This means, not all requests have an expiration. A request that has the phase `""` or `New` will not have its expiration set until it gets processed and returned successfully from the reconcile loop.

Comment (carlisia) [reply pkg/controller/download_request_controller_test.go:85]: Which is to say, only requests that have been successfully processed will have it's expiration set.

File: pkg/controller/download_request_controller_test.go
Comment (carlisia) [reply pkg/controller/download_request_controller_test.go:145]: Ah. I was thinking this test block was a mess. Thanks to your other comment, I realize a bug in the test: a request cannot be both expired and have a phase other than `processed`, and the last test case if braking this rule. Fixing this.

File: pkg/controller/download_request_controller_test.go
Comment (carlisia) [reply pkg/controller/download_request_controller_test.go:262]: See my other two separate comments, they address both points here.

File: pkg/controller/download_request_controller_test.go
Comment (carlisia) [reply pkg/controller/download_request_controller_test.go:85]: Hm. You are right, of course. Nvm what I said. Let me re-evaluate these tests.

**Review by zubron — APPROVED** [review]
Thanks for fixing the tests - LGTM! :)

**Review by nrb — APPROVED** [review]

**Pull request merged**
